### PR TITLE
Segment FBXGeometry and free its unused components

### PR DESCRIPTION
**Pull request by zzmp**
- Split FBXGeometry into parts (SittingPoints, Joints, Meshes, Materials)
- Refactor Animation (Anim*, Rig) to use only joints and animation frames, free unused parts
- Refactor NetworkGeometry to give access to parts, not entire FBXGeometry, free unused parts
- Only copy NetworkMaterial if the texture is explicitly set differently from the cached version
- Refactor EntityItemFBXService to not expose a raw pointer or reference

### CODE REVIEW

**Comment by zzmp** [comment]
See: https://app.asana.com/0/74051501495175/104882336403361.

#### Intent

Eke out a little more memory. Any use of FST/FBX/OBJ holds the entire file, even if it includes parts that it does not use (for example, the scene in demo is loaded twice, once for the model and once for the animation).

#### Testing

Go to a place, let it load completely, and note the memory usage in your task manager.
When I did this, I saw 850MB in current build, 800MB in PR Debug build.

This touches lots of model-related files, so things need to be thoroughly tested to make sure they are still rendering/animating correctly - not just a smoke test.

**Comment by hifi-gustavo** [comment]
Build console output for commit: df4261e3a8555f2528f2dc2d416af9367f44d375 available at: https://gist.github.com/b68c2ca66161ac0cefac9cf5c8afa7d2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7741.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7741.dmg

**Comment by AndrewMeadows** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: df4261e3a8555f2528f2dc2d416af9367f44d375 available at: https://gist.github.com/b68c2ca66161ac0cefac9cf5c8afa7d2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7741.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7741.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 411c75e460db48de9e7f9141c633dec2c10c8404 available at: https://gist.github.com/b68c2ca66161ac0cefac9cf5c8afa7d2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7741.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7741.dmg

**Comment by zzmp** [comment]
I cannot see other people with this PR. DNM.

File: libraries/model-networking/src/model-networking/ModelCache.cpp [hunk 1]
```diff
@@ -265,46 +264,32 @@ std::shared_ptr<NetworkGeometry> ModelCache::getGeometry(const QUrl& url, const
...
-    }
-}
-
 void Geometry::setTextures(const QVariantMap& textureMap) {
     if (_meshes->size() > 0) {
         for (auto& material : _materials) {
             // Check if any material textures actually changed
             if (std::any_of(material->_textures.cbegin(), material->_textures.cend(),
-                [&textureMap](const NetworkMaterial::Textures::value_type& it) { return it.texture && textureMap.contains(it.name); })) { 
+                [&textureMap](const std::vector<NetworkMaterial::Texture>::value_type& it) { return it.texture && textureMap.contains(it.name); })) {
```
Comment (zzmp) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: Does this do what it says?

File: libraries/animation/src/AnimClip.cpp [hunk 2]
```diff
@@ -114,12 +113,13 @@ void AnimClip::copyFromNetworkAnim() {
         jointMap.push_back(skeletonJoint);
     }
 
-    const int frameCount = geom.animationFrames.size();
+    const auto& animationFrames = _networkAnim->getFramesReference();
```
Comment (zzmp) [inline libraries/animation/src/AnimClip.cpp]: !=

File: libraries/animation/src/AnimClip.cpp [hunk 3]
```diff
@@ -114,12 +113,13 @@ void AnimClip::copyFromNetworkAnim() {
...
+    const int frameCount = animationFrames.size();
     _anim.resize(frameCount);
 
     for (int frame = 0; frame < frameCount; frame++) {
 
-        const FBXAnimationFrame& fbxAnimFrame = geom.animationFrames[frame];
+        const auto& fbxAnimFrame = animationFrames[frame];
```
Comment (zzmp) [inline libraries/animation/src/AnimClip.cpp]: s/fbxA/a

File: libraries/animation/src/AnimationCache.cpp [hunk 4]
```diff
@@ -83,7 +83,7 @@ void AnimationReader::run() {
 }
 
 bool Animation::isLoaded() const {
-    return _loaded && _geometry;
+    return _loaded && _joints && _animationFrames;
```
Comment (zzmp) [inline libraries/animation/src/AnimationCache.cpp]: break

File: libraries/animation/src/AnimationCache.h [hunk 5]
```diff
@@ -53,16 +53,14 @@ class Animation : public Resource {
 
     explicit Animation(const QUrl& url);
 
-    const FBXGeometry& getGeometry() const { return *_geometry; }
-
     virtual bool isLoaded() const override;
 
+    const FBXJoints& getJoints() const { return *_joints; }
```
Comment (zzmp) [inline libraries/animation/src/AnimationCache.h]: Move into cpp

File: libraries/animation/src/Rig.cpp [hunk 6]
```diff
@@ -190,18 +190,18 @@ void Rig::initJointStates(const FBXGeometry& geometry, const glm::mat4& modelOff
...
+    _rightHandJointIndex = joints.rightHandJointIndex;
+    _rightElbowJointIndex = _rightHandJointIndex >= 0 ? joints.at(_rightHandJointIndex).parentIndex : -1;
+    _rightShoulderJointIndex = _rightElbowJointIndex >= 0 ? joints.at(_rightElbowJointIndex).parentIndex : -1;
 }
 
-void Rig::reset(const FBXGeometry& geometry) {
-    _geometryOffset = AnimPose(geometry.offset);
-    _animSkeleton = std::make_shared<AnimSkeleton>(geometry);
+void Rig::reset(const FBXJoints& joints) {
+    _geometryOffset = AnimPose(joints.offset);
```
Comment (zzmp) [inline libraries/animation/src/Rig.cpp]: Check offset initialization

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9fd068e1eab605732b0a58a464102828b73d5919 available at: https://gist.github.com/b68c2ca66161ac0cefac9cf5c8afa7d2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7741.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7741.dmg

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 7]
```diff
@@ -226,12 +226,12 @@ EntityItemProperties EntityScriptingInterface::getEntityProperties(QUuid identit
...
-                    const FBXGeometry* geometry = _entityTree->getGeometryForEntity(entity);
-                    if (geometry) {
-                        results.setSittingPoints(geometry->sittingPoints);
-                        Extents meshExtents = geometry->getUnscaledMeshExtents();
-                        results.setNaturalDimensions(meshExtents.maximum - meshExtents.minimum);
-                        results.calculateNaturalPosition(meshExtents.minimum, meshExtents.maximum);
+                    SittingPoints sittingPoints;
+                    Extents extents;
+                    if (_entityTree->getGeometryForEntity(entity, sittingPoints, extents)) {
+                        results.setSittingPoints(sittingPoints);
```
Comment (zzmp) [inline libraries/entities/src/EntityScriptingInterface.cpp]: break

File: libraries/render-utils/src/MeshPartPayload.cpp [hunk 8]
```diff
@@ -305,28 +305,25 @@ ModelMeshPartPayload::ModelMeshPartPayload(Model* model, int _meshIndex, int par
     _shapeID(shapeIndex) {
 
     assert(_model && _model->isLoaded());
-    auto& modelMesh = _model->getGeometry()->getGeometry()->getMeshes().at(_meshIndex);
-    updateMeshPart(modelMesh, partIndex);
+    const auto& mesh = _model->getGeometry()->getMeshes().at(_meshIndex);
```
Comment (zzmp) [inline libraries/render-utils/src/MeshPartPayload.cpp]: !=

**Comment by hifi-gustavo** [comment]
Build console output for commit: d524ec81022568ba3625e7d84b2b3620a3ba0fee available at: https://gist.github.com/b68c2ca66161ac0cefac9cf5c8afa7d2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7741.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7741.dmg

File: libraries/render-utils/src/MeshPartPayload.cpp [hunk 9]
```diff
@@ -377,20 +374,18 @@ ItemKey ModelMeshPartPayload::getKey() const {
...
         return ShapeKey::Builder::invalid();
     }
 
-    const FBXMesh& mesh = geometry.meshes.at(_meshIndex);
+    const auto& mesh = meshes.at(_meshIndex);
 
     // if our index is ever out of range for either meshes or networkMeshes, then skip it, and set our _meshGroupsKnown
     // to false to rebuild out mesh groups.
-    if (_meshIndex < 0 || _meshIndex >= (int)networkMeshes.size() || _meshIndex > geometry.meshes.size()) {
+    if (_meshIndex < 0 || _meshIndex >= (int)meshes.size()) {
```
Comment (zzmp) [inline libraries/render-utils/src/MeshPartPayload.cpp]: !=

File: libraries/render-utils/src/Model.cpp [hunk 10]
```diff
@@ -1086,37 +1088,34 @@ void Model::updateClusterMatrices(glm::vec3 modelPosition, glm::quat modelOrient
...
 bool Model::maybeStartBlender() {
     if (isLoaded()) {
-        const FBXGeometry& fbxGeometry = getFBXGeometry();
-        if (fbxGeometry.hasBlendedMeshes()) {
-            QThreadPool::globalInstance()->start(new Blender(getThisPointer(), ++_blendNumber, _geometry,
-                fbxGeometry.meshes, _blendshapeCoefficients));
+        const auto& geometry = getGeometry();
+        if (geometry->getMeshes().hasBlendedMeshes()) {
+            QThreadPool::globalInstance()->start(
+                new Blender(getThisPointer(), ++_blendNumber, _geometry, _blendshapeCoefficients));
```
Comment (zzmp) [inline libraries/render-utils/src/Model.cpp]: Should this be `getGeometry()`?

File: libraries/render-utils/src/Model.cpp [hunk 11]
```diff
@@ -1182,17 +1181,7 @@ void Model::segregateMeshGroups() {
         assert(isLoaded());
         networkGeometry = _geometry;
     }
-    const FBXGeometry& geometry = networkGeometry->getGeometry()->getGeometry();
-    const auto& networkMeshes = networkGeometry->getGeometry()->getMeshes();
-
-    // all of our mesh vectors must match in size
```
Comment (zzmp) [inline libraries/render-utils/src/Model.cpp]: !=

File: libraries/render-utils/src/Model.h [hunk 12]
```diff
@@ -97,7 +97,7 @@ class Model : public QObject, public std::enable_shared_from_this<Model> {
                     bool showCollisionHull = false);
     void removeFromScene(std::shared_ptr<render::Scene> scene, render::PendingChanges& pendingChanges);
     void renderSetup(RenderArgs* args);
-    bool isRenderable() const { return !_meshStates.isEmpty() || (isActive() && getGeometry()->getGeometry()->getMeshes().empty()); }
+    bool isRenderable() const { return !_meshStates.isEmpty() || (isActive() && !getGeometry()->getMeshes().isEmpty()); }
```
Comment (zzmp) [inline libraries/render-utils/src/Model.h]: !=

**Comment by hifi-gustavo** [comment]
Build console output for commit: afa3568354f816633cb3323642e51ef615272904 available at: https://gist.github.com/b68c2ca66161ac0cefac9cf5c8afa7d2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7741.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7741.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6bc9986249a5bd217b2fa4895a6b1a7294856991 available at: https://gist.github.com/b68c2ca66161ac0cefac9cf5c8afa7d2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7741.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7741.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8449e1366b99b16d056ac07453cad9e7f86bf6de available at: https://gist.github.com/b68c2ca66161ac0cefac9cf5c8afa7d2<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7741.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7741.dmg

**Comment by zzmp** [comment]
Stale. Closing - will rework and reopen when it is ready.

**Pull request closed without merging**